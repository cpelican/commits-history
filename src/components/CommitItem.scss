:root {
    --box-shadow-color: rgba(0, 0, 0, 0.556);
}

.commit-item {
    width: 400px;
    margin: 1rem;
    padding: 1rem;
    border-radius: 8px;
    list-style-type: none;
    box-shadow: 0px 0px 20px 4px var(--box-shadow-color);

    h4 {
        margin-top: 0;
        margin-bottom: .2rem;
        &::first-letter {
            text-transform: uppercase;
        }
    }

    .additional-info-container {
        margin-top: .5rem;

        .author-message {
            &:after {
                content: '\2022';
                margin: .2rem;
            }
        }
    }
}
