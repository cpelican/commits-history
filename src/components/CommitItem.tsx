import React from 'react';
import {DateTime} from 'luxon';

import './CommitItem.scss';
import {AdditionalInfo} from './AdditionalInfo';
import {CommitTypeBadge, commitTypes} from './CommitTypeBadge';

export interface CommitItemProps {
    sha: string;
    date: string;
    message: string;
    author: {
        name: string;
        imageUrl: string;
    };
}

export interface CommitBits {
    commitMessage: string;
    prefix?: keyof typeof commitTypes;
}

export function getMessageBits(commitMessage: string): CommitBits {
    if (commitMessage == null) {
        return {commitMessage: ''};
    }
    // eslint-disable-next-line no-unsafe-optional-chaining
    const [commitTypePrefix, ...rest] = commitMessage?.trim?.()?.split(':'),
        commitType = commitTypes?.[commitTypePrefix];
    if (commitType == null) {
        return {commitMessage};
    }
    const splittedMessage = rest?.join(':')?.trim();
    return {prefix: commitType, commitMessage: splittedMessage === '' ? commitType : splittedMessage};
}

export const CommitItem: React.FC<CommitItemProps> = React.memo(({sha, message, date, author}) => {
    function getDateMessage() {
        const dateMessage = date ? DateTime.fromISO(date).toRelative({unit: ['years', 'months', 'days', 'hours', 'minutes']}) : '-';
        return <span>{dateMessage}</span>;
    }

    function getAuthor() {
        const name = author.name ?? 'unknown';
        return <span className='author-message'>{name}</span>;
    }

    const content = <>{getAuthor()}{getDateMessage()}</>,
        {prefix, commitMessage} = getMessageBits(message);
    return (
        <li key={sha} className='commit-item'>
            <h4>{commitMessage === '' ? 'No commit message' : commitMessage}</h4>
            <CommitTypeBadge type={prefix}/>
            <AdditionalInfo content={content} imageUrl={author.imageUrl} imageAlt={`avatar for ${author.name}`}/>
        </li>
    );
});

CommitItem.displayName = 'CommitItem';
