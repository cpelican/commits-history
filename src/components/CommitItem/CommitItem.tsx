import React from 'react';
import {DateTime} from 'luxon';

import './CommitItem.scss';
import {AdditionalInfo} from '../AdditionalInfo';
import {CommitTypeBadge} from '../CommitTypeBadge';
import {getMessageBits} from '../helpers';

export interface CommitItemProps {
    sha: string;
    date: string;
    message: string;
    author: {
        name: string;
        imageUrl: string;
    };
}

export const CommitItem: React.FC<CommitItemProps> = React.memo(({sha, message, date, author}) => {
    function getDateMessage() {
        const dateMessage = date
            ? DateTime.fromISO(date).toRelative({unit: ['years', 'months', 'days', 'hours', 'minutes']})
            : '-';
        return <span>{dateMessage}</span>;
    }

    function getAuthor() {
        const name = author.name ?? 'unknown';
        return <span className='author-message'>{name}</span>;
    }

    const content = (
            <>
                {getAuthor()}
                {getDateMessage()}
            </>
        ),
        {prefix, commitMessage} = getMessageBits(message);
    return (
        <li key={sha} className='commit-item'>
            <h4>{commitMessage === '' ? 'No commit message' : commitMessage}</h4>
            <CommitTypeBadge type={prefix} />
            <AdditionalInfo imageUrl={author.imageUrl} imageAlt={`avatar for ${author.name}`} >
                {content}
            </AdditionalInfo>
        </li>
    );
});

CommitItem.displayName = 'CommitItem';
