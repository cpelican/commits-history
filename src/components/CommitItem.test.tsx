import React from 'react';
import {CommitBits, CommitItem, CommitItemProps, getMessageBits} from './CommitItem';
import {render, screen} from '@testing-library/react';

const defaultProps: CommitItemProps = {
    sha: '1234',
    date: 'one minute ago',
    message: 'commit message',
    author: {
        imageUrl: 'www.avatar.com',
        name: 'some name',
    },
};

describe('CommitItem', () => {
    test.each<[string, CommitBits]>([
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        [null, {commitMessage: ''}],
        ['', {commitMessage: ''}],
        ['test', {commitMessage: 'test', prefix: 'test'}],
        ['add tests', {commitMessage: 'add tests'}],
        ['wtf message: no no no!', {commitMessage: 'wtf message: no no no!'}],
        ['many: double: points', {commitMessage: 'many: double: points'}],
        ['chore: do something', {commitMessage: 'do something', prefix: 'chore'}],
        ['chores: do something', {commitMessage: 'chores: do something'}],
        ['feature:do something', {commitMessage: 'do something', prefix: 'feature'}],
        [
            'feature:do something: do another thing',
            {commitMessage: 'do something: do another thing', prefix: 'feature'},
        ],
        [
            'feature: do something: do another thing',
            {commitMessage: 'do something: do another thing', prefix: 'feature'},
        ],
    ])('getMessageBits() with commit message %s renders %s', (commitMessage, expected) => {
        expect(getMessageBits(commitMessage)).toEqual(expected);
    });

    test('renders as expected', () => {
        const {asFragment} = render(<CommitItem {...defaultProps} />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('renders as expected with no message', () => {
        render(<CommitItem {...{...defaultProps, message: ''}} />);
        expect(screen.getByRole('heading', {level: 4})).toMatchSnapshot();
    });
});
