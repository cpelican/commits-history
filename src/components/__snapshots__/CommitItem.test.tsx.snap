// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommitItem renders as expected 1`] = `
<DocumentFragment>
  <li
    class="commit-item"
  >
    <h4>
      commit message
    </h4>
    <div
      class="additional-info-container"
    >
      <img
        alt="avatar for some name"
        class="additional-info-image"
        src="www.avatar.com"
      />
      <div
        class="additional-info-content"
      >
        <span
          class="author-message"
        >
          some name
        </span>
        <span />
      </div>
    </div>
  </li>
</DocumentFragment>
`;

exports[`CommitItem renders as expected with no message 1`] = `
<h4>
  No commit message
</h4>
`;
